### DISCUSSION TITLE
Enhancement: Added diagnostic check for a call expression that appears within a type annotation

### DISCUSSION THREAD

**Discussion by omerc7** [microsoft/pyright#1594] [General]
Hi,
While working with Pydantic I noticed errors raised by Pylance 2021.3.0 that where not raised before, error:
"Illegal type annotation: call expression not allowed"
![Screen Shot 2021-03-08 at 11 30 15](https://user-images.githubusercontent.com/32813109/110302258-b8484a00-8001-11eb-8189-b9bae85f08a0.png)
* a screen shot of previous version below

A look at Pylance's change log reviled an upgrade of Pyright from 1.1.114 to 1.1.117.
Pyright release notes on version 1.1.116 [(link)](https://github.com/microsoft/pyright/releases/tag/1.1.116) includes: 

"Enhancement: Added diagnostic check for a call expression that appears within a type annotation. This was previously not flagged as an error."

I'm wondering where this came from and if it's considered 'bad practice' to have a call expression that returns a type annotation. I could not find anything about it in PEP484.

As FastAPI and with it the use of Pydantic is growing rapidly, I'm questioning if this is an enhancement or not, making Pyright not compatible with "Pydantic's type system".

I would love to hear thoughts about this..

This print screen if from Pylance 2021.2.4 using Pyright 1.1.114 of the same code as above. Inferring the right types
![Screen Shot 2021-03-08 at 11 47 45](https://user-images.githubusercontent.com/32813109/110304523-6228d600-8004-11eb-9baf-e883769198c8.png)

**Comment by erictraut** [comment]
Here's what PEP 484 says:

>Annotations must be valid expressions that evaluate without raising exceptions at the time the function is defined (but see below for forward references).

>Annotations should be kept simple or static analysis tools may not be able to interpret the values. For example, dynamically computed types are unlikely to be understood. (This is an intentionally somewhat vague requirement, specific inclusions and exclusions may be added to future versions of this PEP as warranted by the discussion.)

A type annotation expression need to be specified in a form that is static, simple, and unambiguous regardless of the type checker used. It needs to be evaluated without any type inference, overload matching, constraint solving, etc. because those behaviors can vary by type checker. Those restrictions imply that call expressions cannot be used here.

Mypy also flags function calls within type annotations as an error, so this issue isn't specific to pyright. It was simply an oversight that pyright didn't previously generate an error in this case.

PEP 484 clearly lays out support for type aliases, which can be used as a simple form of "macro" if you have a complex type that you want to parameterize. The use of type aliases are fully supported in type annotations.

**Reply by erictraut** [reply]
If you want to use type annotations for something other than static type analysis (e.g. for dynamic type analysis or enforcement), perhaps those other annotations should use the [PEP 593](https://www.python.org/dev/peps/pep-0593/) facility, which does support arbitrary expressions.

**Reply by omerc7** [reply]
That makes sense..
Thanks for the quick and thorough response, much appreciated

**Reply by mmulet** [reply]
@erictraut I agree with you, but at the end of the day this is a regression of previously supported uses cases. In my example, I'm developing a plugin for [blender](https://www.blender.org/), and the [blender api](https://docs.blender.org/api/current/bpy.props.html) (wisely or not) uses call expressions to create metadata for its properties. Because of this change, I have remove all type checking from my program or keep pyright fixed to version 1.1.115 for all time.  
Any chance of separating this type check into a new flag so I can use # pyright: reportIllegalCallExpression=false or similar?

xkcd: 1172
![xkcd](https://imgs.xkcd.com/comics/workflow.png)

**Reply by erictraut** [reply]
Sorry, but call expressions were not meant to be supported in type annotations. The fact that they happened to work in some situations was an oversight, not a feature. I feel bad that my oversight is now creating extra work for you, but perpetuating the problem is not a good solution.

**Reply by erictraut** [reply]
It honestly never occurred to me that someone would try to use a call expression in a type annotation, so I didn't think to add an explicit check for it. It wasn't until someone pointed out to me that mypy generates an explicit error that I realized pyright should also emit an error in this case.

**Comment by mmulet** [comment]
PEP 484 doesn't define must or should, but most technical specification follows [rfc2119](https://tools.ietf.org/html/rfc2119) which defines MUST as "REQUIRED" and SHOULD as "RECOMMENDED". So, it is only Recommended, but not Required that annotations be kept simple. So, even with call expressions intact, pyright would still be conformant to the spec.
The fact that it's only a recommendation allows programs like blender or pydantic to use call expressions as types in the first place. 
Last but not least, the spec says that type expressions should be simple, so as to not confuse the static analyzer. It's slightly backwards to forbid complex expressions in the static analyzer to avoid confusing itself.

 I don't know mypy's reasoning for disallowing the call expressions, I'll probably ask them later.

**Reply by erictraut** [reply]
Type annotations need to be unambiguous in their interpretation. The result should not differ between type checkers. There are subtle differences in the type results for call expressions. That fact alone is a highly compelling reason to disallow call expressions within annotations.

Also, call expressions are relatively expensive to analyze, and it's important for type annotations to be relatively cheap because they are on the critical path for presenting completion suggestions during interactive editing.

There's already a mechanism defined in PEP 484 for something akin to call expressions: type aliases. These are simpler than calls (e.g. they don't allow overloads), so they don't run afoul of the problems of general call expressions. Yet they allow for simple parameterization and even generic type specialization.

I don't know the details of how you're using call expressions in type annotations or why you decided that was desirable in your particular case, but I wonder whether it's feasible to replace them with type aliases instead. That sounds like a relatively straightforward replacement, and it would work with all existing type checkers.

**Reply by mmulet** [reply]
If I were writing my own code-base I would definitely use Type alias, but I am not. So I don't have much choice when it comes to api. This is what I'm working with:
```python
# bpy.props stubs file
from typing import Type
# most fields are omitted for this example
def StringProperty(name: str =...) -> Type[str]: ...

# File.py
from bpy.types import Panel
from bpy.props import StringProperty

class SomeUIPanel(Panel):
   some_string_prop: StringProperty(name="User facing string property name")
```
At runtime, blender uses the StringProperty to encode metadata onto the some_string_prop, that they query at runtime to build the UI.
Blender throws an error if you use
```python
class SomeUIPanel(Panel):
   # notice the = sign instead of :
   some_string_prop = StringProperty(name="User facing string property name")
```
So it is clear that type annotation is the intended behavior. They also use it in their documentation [as you can see here](https://docs.blender.org/api/current/bpy.props.html#operator-example)

### How about a compromise?
There doesn't have to subtle differences in type results of call expression if they defined like this: 
```
# notice the explicit Type[str] return type
def StringProperty(name: str =...) -> Type[str]: ...
```
How about allowing call expression types if they have an explicit return Type extending form typing.Type? This will eliminate differences, and be performant.

**Reply by erictraut** [reply]
We are trying to drive consistency across type checkers in the Python community with respect to how libraries annotate their types. It has been a bit of a wild west up until now, and everyone in the community has suffered because of this.

If the problem is in the blender API library, then the maintainers of that library will need to conform to the typing standards. Please file an issue with them and point them to this discussion plus [this documentation](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md). Or if you feel so inclined, submit a PR that fixes the issue.

**Comment by adriangb** [comment]
Since Pydantic was brought up in the original comment I wanted to follow up on the current situation for Pydantic.

Pydantic now supports PEP593 so instead of the original example one can write:

```python
from typing import Annotated
from pydantic import BaeModel, Field

class Model(BaseModel):
    bit_int: Annotated[int, Field(gt=1000, lt=1024)]
```

While this works, we still have two issues.

###  Backwards compatibility

This is a problem specific to Pydantic and not wanting to make a breaking change. It's not reason enough to make changes in a type checker but worth pointing out.

### Verbosity

We now have to force users to always import and type out `Annotated` and `Field`, which gets very tiring and repetitive since it's not uncommon to have hundreds of fields across models.
We also have to teach users about PEP593 or at least it's usage for them to understand _why_ they have to do things this way.

It would be very nice indeed if this would work:

```python
def conint(gt: int | None = None) -> type[int]:
    return Annotated[int, Field(gt=gt)]

class Model(BaseModel):
    big_int: conint(gt=1000)
```

This was already suggested in https://github.com/microsoft/pyright/discussions/1594#discussioncomment-456747 and is [precisely what Pydantic will be doing going forward](https://github.com/pydantic/pydantic/blob/ef8f9fe51b9d0e38693f70482c29ba124f810d5a/pydantic/types.py#L96-L110).

I think supporting this would make PEP 593 much easier to work with for many libraries because they can hide that complexity from users and would make a lot of users of Pyright/Pydantic much happier.

Since this is a very specific use case where the type can be obtained from the return type annotation, would it be possible to support this without too much of a performance penalty or complexity for the type checker? I'm open to trying my hand at an implementation but since I'm not a TS dev it's going to be slow; I want to make sure there's some buy in before I burn the time.

**Reply by erictraut** [reply]
Call expressions are not supported in type annotation expressions, nor should they be. Type annotation intended to be fast to evaluate and unambiguous in their meaning. Call expressions are expensive to evaluate and are subject to under-specified overload rules, constraint solving heuristics, etc. So no, this isn't something we would consider supporting in pyright.

**Reply by adriangb** [reply]
Thanks for the quick reply.

I was hoping that narrowing the use case to call expressions that explicitly return a type (maybe even disallowing overloads or anything else that would cause complexity) would help keep things simple and performant but it sounds like this is a door you don't want to open at all, which I understand. It is unfortunate because I think that using call expressions to simplify usage of PEP 593 would be a very popular use case and would eliminate the need for a lot of weird things libraries do to cram information into type annotations.

**Reply by erictraut** [reply]
There is a strong consensus among Python type checker authors that call expressions should not be supported in type annotation expressions. There are solid reasons for this. So yes, you should assume that you will receive strong pushback from all static type checkers with such a proposal.

**Reply by adriangb** [reply]
Fair enough, thank you.
